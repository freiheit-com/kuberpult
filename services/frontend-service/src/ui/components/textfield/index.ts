import { Textfield } from './textfield';

export { Textfield };
