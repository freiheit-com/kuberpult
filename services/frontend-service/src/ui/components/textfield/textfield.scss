@use '@material/floating-label/mdc-floating-label';
@use '@material/notched-outline/mdc-notched-outline';
@use '@material/textfield/mdc-text-field';
@use '@material/textfield/icon';

@include icon.icon-core-styles;

.mdc-text-field {
  width: 100%;
}

div.mdc-text-field.mdc-text-field--outlined span.mdc-floating-label.mdc-floating-label--float-above {
	@extend .text-bold;
}

span.mdc-floating-label {
	@extend .text-bold;
}
