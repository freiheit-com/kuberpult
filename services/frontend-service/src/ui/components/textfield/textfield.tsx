/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with kuberpult.  If not, see <http://www.gnu.org/licenses/>.

Copyright 2021 freiheit.com*/
import { useCallback, useEffect, useRef } from 'react';
import classNames from 'classnames';
import { MDCTextField } from '@material/textfield';
import { useSearchParams } from 'react-router-dom';

export type TextfieldProps = {
    className?: string;
    floatingLabel?: string;
    value?: string | number;
    leadingIcon?: string;
};

export const Textfield = (props: TextfieldProps) => {
    const { className, floatingLabel, leadingIcon, value } = props;
    const control = useRef<HTMLDivElement>(null);
    const input = useRef<HTMLInputElement>(null);
    const MDComponent = useRef<MDCTextField>();

    useEffect(() => {
        if (control.current) {
            MDComponent.current = new MDCTextField(control.current);
        }
        return () => MDComponent.current?.destroy();
    }, []);

    useEffect(() => {
        if (floatingLabel) {
            MDComponent.current?.layout();
        }
    });

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [searchParams, setSearchParams] = useSearchParams();

    const allClassName = classNames(
        'mdc-text-field',
        'mdc-text-field--outlined',
        {
            'mdc-text-field--no-label': !floatingLabel,
            'mdc-text-field--with-leading-icon': leadingIcon,
        },
        className
    );

    const setQueryParam = useCallback(
        (event: any) => {
            if (event.target.value !== '') searchParams.set('application', event.target.value);
            else searchParams.delete('application');
            setSearchParams(searchParams);
        },
        [searchParams, setSearchParams]
    );

    return (
        <div className={allClassName} ref={control}>
            <span className="mdc-notched-outline">
                <span className="mdc-notched-outline__leading" />
                {!!floatingLabel && (
                    <span className="mdc-notched-outline__notch">
                        <span
                            className={classNames('mdc-floating-label', {
                                'mdc-floating-label--float-above':
                                    !!value ||
                                    (input.current && input.current.value !== '') ||
                                    input.current === document.activeElement,
                            })}>
                            {floatingLabel}
                        </span>
                    </span>
                )}
                <span className="mdc-notched-outline__trailing" />
            </span>
            {leadingIcon && (
                <i className="material-icons mdc-text-field__icon mdc-text-field__icon--leading" tabIndex={0}>
                    {leadingIcon}
                </i>
            )}
            <input
                type="text"
                className="mdc-text-field__input"
                defaultValue={value}
                ref={input}
                aria-label={floatingLabel}
                disabled={window.location.href.includes('environments')}
                onChange={setQueryParam}
            />
        </div>
    );
};
