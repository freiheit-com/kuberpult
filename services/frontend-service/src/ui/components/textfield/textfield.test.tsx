/*This file is part of kuberpult.

Kuberpult is free software: you can redistribute it and/or modify
it under the terms of the Expat(MIT) License as published by
the Free Software Foundation.

Kuberpult is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
MIT License for more details.

You should have received a copy of the MIT License
along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.

Copyright 2023 freiheit.com*/
import { Textfield } from './textfield';
import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

describe('Textfield', () => {
    it('renders correctly using Snapshot', () => {
        // given
        const { container } = render(
            <MemoryRouter>
                <Textfield floatingLabel="Floating label" />
            </MemoryRouter>
        );
        // when & then
        expect(container.firstChild).toMatchSnapshot();
    });

    test('renders correctly with leading icon', () => {
        // given
        const { container } = render(
            <MemoryRouter>
                <Textfield leadingIcon="search" />
            </MemoryRouter>
        );
        // when & then
        expect(container.querySelectorAll('div')[0]?.className).toEqual(
            'mdc-text-field mdc-text-field--outlined mdc-text-field--no-label mdc-text-field--with-leading-icon'
        );
        expect(container.querySelector('i')).toMatchInlineSnapshot(`
    <i
      class="material-icons mdc-text-field__icon mdc-text-field__icon--leading"
      role="button"
      tabindex="0"
    >
      search
    </i>
  `);
    });
});

describe('Verify textfield content', () => {
    const getNode = (overrides?: {}): JSX.Element | any => {
        // given
        const defaultProps: any = {
            children: null,
        };
        return (
            <MemoryRouter>
                <Textfield {...defaultProps} {...overrides} />;
            </MemoryRouter>
        );
    };

    const getWrapper = (overrides?: { floatingLabel: string; value: string; className: string }, entries?: string[]) =>
        render(getNode(overrides));

    it(`Renders a navigation item base`, () => {
        // when
        const { container } = getWrapper();
        // then
        expect(container.firstChild);
    });

    interface dataT {
        name: string;
        className: string;
        floatingLabel: string;
        value: string;
        expect: (container: HTMLElement) => void;
    }

    const data: dataT[] = [
        {
            name: 'Empty textfield',
            className: 'top-app-bar-search-field',
            floatingLabel: 'Search',
            value: '',
            expect: (container) =>
                expect(container.getElementsByClassName('mdc-text-field__input')[0]).toHaveTextContent(''),
        },
        {
            name: 'Textfield with content',
            className: 'top-app-bar-search-field',
            floatingLabel: 'Search',
            value: 'test-search',
            expect: (container) => {
                // eslint-disable-next-line no-type-assertion/no-type-assertion
                const input = container.getElementsByClassName('mdc-text-field__input')[0] as HTMLElement;
                input.nodeValue = 'test-search';
                return expect(container.getElementsByClassName('mdc-text-field__input')[0]).toHaveDisplayValue(
                    'test-search'
                );
            },
        },
    ];

    describe.each(data)(`Renders a navigation item with selected`, (testcase) => {
        it(testcase.name, () => {
            const { className, floatingLabel, value } = testcase;
            // when
            const { container } = getWrapper({ floatingLabel, className, value });
            // then
            testcase.expect(container);
        });
    });
});
