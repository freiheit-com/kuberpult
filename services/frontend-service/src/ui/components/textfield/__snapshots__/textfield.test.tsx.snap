// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textfield renders correctly using Snapshot 1`] = `
<div
  class="mdc-text-field mdc-text-field--outlined mdc-text-field--no-label"
>
  <span
    class="mdc-notched-outline"
  >
    <span
      class="mdc-notched-outline__leading"
    />
    <span
      class="mdc-notched-outline__trailing"
    />
  </span>
  <input
    aria-label="Floating label"
    class="mdc-text-field__input"
    id="application-filter-input"
    placeholder="Floating label"
    type="search"
    value=""
  />
</div>
`;
